pkg/proxy/topology.go

```diff
@@ -160,19 +164,23 @@ func canUseTopology(endpoints []Endpoint, svcInfo ServicePort, nodeLabels map[st
 		}
 	}
 
-	zone, ok := nodeLabels[v1.LabelTopologyZone]
-	if !ok || zone == "" {
-		klog.V(2).InfoS("Skipping topology aware endpoint filtering since node is missing label", "label", v1.LabelTopologyZone)
-		return false
-	}
-
+	zone, foundZone := nodeLabels[v1.LabelTopologyZone]
 	hasEndpointForZone := false
 	for _, endpoint := range endpoints {
 		if !endpoint.IsReady() {
 			continue
 		}
+
+		// If any of the endpoints do not have zone hints, we bail out
 		if endpoint.ZoneHints().Len() == 0 {
-			klog.V(2).InfoS("Skipping topology aware endpoint filtering since one or more endpoints is missing a zone hint", "endpoint", endpoint)
+			return false
+		}
+
+		// If there is at least one zone hint, we check if there is a zone label, if there isn't one we bail out
+		if !foundZone || zone == "" {
+			logOnce.Do(func() {
```

Comment:
I may be missing something, but I don't think this is needed if we're always going to return after this log message.

Reply:
since the `logOnce` is global we log this line just once regardless of how many times `canUseTopology` is called.

Reply:
Yeah that is not something that we would want, i.e. only logging this once (if at all) for the lifetime of the process.

Sorry for making this more involved for you, but is it better if we instead have something like this:

```
	hasEndpointForZone := false
	endpointsWithHint = 0
	for _, endpoint := range endpoints {
		if !endpoint.IsReady() {
			continue
		}
		if endpoint.ZoneHints().Len() == 0 {
			klog.V(2).InfoS("Skipping topology aware endpoint filtering since one or more endpoints is missing a zone hint", "endpoint", endpoint)
			return false
		}
		endpointsWithHint++
		if endpoint.ZoneHints().Has(zone) {
			hasEndpointForZone = true
		}
	}

	allEndpointsWithHint := endpointsWithHint == len(endpoints)
	if allEndpointsWithHint && (!foundZone || zone == ") {
		klog.V(2).InfoS("Skipping topology aware endpoint filtering since node is missing label", "label", v1.LabelTopologyZone)
	}
```

There's obvious cons that it doesn't shortcircuit earlier, but like @robscott suggested, this should only get triggered for real errors (and be almost very rare, unless there's some real bug)

Reply:
@gauravkghildiyal sorry i lost you ... should i keep the code as-is or do you want me to remove `logOnce`?

Reply:
Went down a bit of a rabbit hole here, but since kube-proxy can now respond to changes on Nodes (included zone labels), it is valuable to log this whenever it happens, and not just the first time. So after some discussion with @gauravkghildiyal, we think that the logOnce wrapper should be removed here.

This will still only log if a node has no zone _and_ an EndpointSlice has a hint, which means something has gone wrong. The EndpointSlice controller will not set hints if a node does not have a zone label, so there's something else (ie custom EndpointSlices) that are causing a problem. I think this change should still remove >99.9% of the logs here.

Reply:
Done. (removed `logOnce`)